xdsl/transforms/memref_stream_tile_outer_loops.py

```diff
@@ -0,0 +1,216 @@
+from collections.abc import Sequence
+from dataclasses import dataclass, field
+from typing import cast
+
+from xdsl.context import MLContext
+from xdsl.dialects import affine, arith, memref, memref_stream, scf
+from xdsl.dialects.builtin import (
+    AffineMapAttr,
+    ArrayAttr,
+    IndexType,
+    IntegerAttr,
+    MemRefType,
+    ModuleOp,
+    NoneAttr,
+    StridedLayoutAttr,
+)
+from xdsl.ir import Attribute, Block, Operation, Region, SSAValue
+from xdsl.ir.affine import AffineMap
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint, Rewriter
+from xdsl.utils.exceptions import DiagnosticException
+
+
+def insert_subview(
+    memref_val: SSAValue,
+    affine_map: AffineMap,
+    apply_operands: Sequence[SSAValue],
+    upper_bounds: Sequence[int],
+    location: InsertPoint,
+):
+    """
+    A helper method to insert a subview from the input `memref_val` to one with new upper
+    bounds, given that it will be accessed with the specified affine map.
+    `apply_operands` are the operands to use to determine the new offset,
+    and `upper_bounds` the new shape.
+    Any new operations should be inserted at `location`.
+    """
+    name_hint_prefix = (
+        memref_val.name_hint + "_" if memref_val.name_hint is not None else ""
+    )
+    apply_ops = tuple(
+        affine.ApplyOp(
+            apply_operands,
+            AffineMapAttr(
+                AffineMap(affine_map.num_dims, affine_map.num_symbols, (res,))
+            ),
+        )
+        for res in affine_map.results
+    )
+    offset_vals = tuple(apply_op.result for apply_op in apply_ops)
+    for offset_val in offset_vals:
+        offset_val.name_hint = name_hint_prefix + "offset"
+    Rewriter.insert_op(apply_ops, location)
+
+    # New subview shape
+    max_values = tuple(ub - 1 for ub in upper_bounds)
+    sizes = tuple(
+        apply_op.map.data.eval(max_values, ())[0] + 1 for apply_op in apply_ops
+    )
+
+    source_type = memref_val.type
+    if not isinstance(source_type, memref.MemRefType):
+        raise DiagnosticException("Cannot create subview from non-memref type")
```

Comment:
(would be cool to have a typed SSA value in xDSL to move such checks to `pyright`)

Reply:
I agree :)